text/000-rustdoc-jump-to-definition.md

```diff
@@ -0,0 +1,352 @@
+
+Another important note is that if documentation is split in multiple places or requires an external tool for the source code navigation, it's actually slowing down our users quite a lot.
+
+A good example where this would be very helpful is for the [rust compiler source code](https://doc.rust-lang.org/1.57.0/nightly-rustc/src/rustc_middle/hir/map/mod.rs.html#872-877): even if a fully set up IDE or github, it's very hard to go around without previous knowledge or help from someone else. It's also very common to have undocumented items or to just want to see how something is implemented. With the "jump to definition", it already improved quite a lot the browsing experience by allowing to jump super quickly by simply clicking on a item. The missing part is now to go back from the source code into the item documentation to make it complete.
+
+Here is a video showing the feature in action:
+
+https://user-images.githubusercontent.com/3050060/114622354-21307b00-9cae-11eb-834d-f6d8178a37bd.mp4
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This feature **only** targets the source code pages.
+
+It can be implemented without JavaScript additions to Rustdoc as pure link generation; so it will work without JS enabled. Rustdoc already has an internal link generation system that can be used here.
+
+## Link generation
+
+There are two kinds of links:
+
+ * Jump to documentation.
+ * Jump to definition.
+
+### Jump to documentation
+
+All items with a documentation page should have a link to it (like methods, trait/impl associated items, etc). However, fields, variants and impl blocks should not (explications below).
+
+The link is generated on their ident/name where they are declared, not when they are in use. For example:
+
+```rust
+pub struct Foo;
+
+let x = Foo;
+```
+
+On `Foo` definition (`pub struct Foo`), we generate a link to its documentation page whereas on its usage (`x = Foo`), we generate a link to jump to its definition.
+
+This is why generating a link for the impl blocks would be complicated: on `impl Foo`, we generate a "jump to definition" link on `Foo`, leaving no space to generate a link to the impl documentation.
+
+For fields and variants, it's simply to limit the quantity of links generated: you can simply click on their parent item to get to their documentation page.
+
+### Jump to definition
+
+A small preamble before listing the items we should generate links for: if an item which should get a link is preceded by a path, the whole path is used as a link. For example:
+
+```rust
+let x: b::c::D; // `b::c::D` is the link to `D`.
+foo::some_fn(); // `foo::some_fn` is the link to `some_fn`.
+```
+
+Anything not specifically mentioned should not get a link generated for itself.
+
+#### Local variables
+
+Local variables should get a link to their definition (not their type):
```

Comment:
Oh I see. I originally fixed this by hiding the compiler errors because they don't impact the final output although errors are emitted. It was reverted afterward so they're displayed again. I'm too sure what to do about this.

Reply:
Absolutely. I just wanted to know first if we wanted to propose a solution in the RFC or not. I plan to add this and the "up limit" to the RFC.

Reply:
When you're not sure what you should do about a problem, you should write that in the RFC. In particular, you should write:

 - what the problem is
 - what its impacts are (for instance, does this cause all crates to fail to build with this flag? only some crates?)
 - potentially some ideas about how to fix it, with the tradeoffs for those ideas

But if you don't have ideas for how to fix it, it's still important to write down the problem so that other people can help you figure out solutions (and so they don't have to discover the problem themselves from scratch).